ice/compat: Remove support for date -> timestamp promotion

oleiman opened the pull request:
This is a v3 feature. It's easy to show that promoting fields this way breaks both Spark and Trino.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [x] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

* none

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

oleiman committed:
ice/compat: Remove support for date -> timestamp promotion

This is a v3 feature. It's easy to show that promoting fields this way breaks
both Spark and Trino.

Signed-off-by: Oren Leiman <oren.leiman@redpanda.com>

rockwotj reviewed (dismissed):
Do we need to make sure there is `SELECT *` in all our schema evolution tests?

oleiman commented:
> Do we need to make sure there is `SELECT *` in all our schema evolution tests?

Yes we do, but this was the only case that didn't have one. Score one for "untested code is broken code".

oleiman commented:
Oh, actually I suppose we're only selecting fields of interest in these tests. Probably a blanket `SELECT *` would be advisable, as a smoke test.

oleiman committed:
dt/dl: Perform a 'SELECT *' query after every schema evolution test

The previous commit fixes a bug that would have been caught if there had been
a SELECT query in the corresponding ducktape test. It's probably a good idea to
fire this off at the end of every test, as a rule.

oleiman commented:
/ci-repeat 1

oleiman commented:
/ci-repeat 1

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--d383085c-62d7-4e17-8992-14f442f8c9cd-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/62708">build#62708</a></summary>


| test_id                                                                                                                                                                                                                                            | test_kind   | job_url                                                                                   | test_status   | passed   |
|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| kafka_server_rpfixture.kafka_server_rpfixture                                                                                                                                                                                                      | unit        | https://buildkite.com/redpanda/redpanda/builds/62708#019569c0-01f8-48ab-ad84-6ab51711d469 | FLAKY         | 1/2      |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_dropped_column_no_collision.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.produce_mode=ProducerType.AVRO.catalog_type=CatalogType.NESSIE      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62708#01956a1b-fc11-4b08-b588-3079a4ea1726 | FLAKY         | 1/3      |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_dropped_column_no_collision.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.produce_mode=ProducerType.AVRO.catalog_type=CatalogType.REST_JDBC   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62708#01956a1b-fc0f-41c6-aafc-7e0e7f81756e | FLAKY         | 1/2      |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_dropped_column_no_collision.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.produce_mode=ProducerType.PROTO2.catalog_type=CatalogType.NESSIE    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62708#01956a1b-fc10-40a2-ae41-0adba94d14bf | FLAKY         | 1/4      |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_dropped_column_no_collision.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.SPARK.produce_mode=ProducerType.PROTO2.catalog_type=CatalogType.NESSIE    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62708#01956a1e-f1f4-4c59-b704-4cec7e4cfdc2 | FLAKY         | 1/8      |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_dropped_column_no_collision.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.produce_mode=ProducerType.AVRO.catalog_type=CatalogType.NESSIE      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62708#01956a1b-fc0e-4fb4-8a2d-c0dd68f992ca | FLAKY         | 1/18     |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_dropped_column_no_collision.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.produce_mode=ProducerType.AVRO.catalog_type=CatalogType.NESSIE      | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62708#01956a1e-f1f6-4c2c-996f-679eac581f9d | FLAKY         | 1/7      |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_dropped_column_no_collision.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.produce_mode=ProducerType.AVRO.catalog_type=CatalogType.REST_JDBC   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62708#01956a1b-fc10-40a2-ae41-0adba94d14bf | FLAKY         | 1/3      |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_dropped_column_no_collision.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.produce_mode=ProducerType.PROTO2.catalog_type=CatalogType.NESSIE    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62708#01956a1b-fc11-4b08-b588-3079a4ea1726 | FLAKY         | 1/8      |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_dropped_column_no_collision.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.produce_mode=ProducerType.PROTO2.catalog_type=CatalogType.NESSIE    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62708#01956a1e-f1f5-45c3-b758-e7b1ff4ea8a8 | FLAKY         | 1/11     |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_dropped_column_no_collision.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.produce_mode=ProducerType.PROTO2.catalog_type=CatalogType.REST_JDBC | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62708#01956a1b-fc0f-41c6-aafc-7e0e7f81756e | FLAKY         | 1/2      |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_dropped_column_no_collision.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.produce_mode=ProducerType.PROTO3.catalog_type=CatalogType.NESSIE    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62708#01956a1b-fc10-40a2-ae41-0adba94d14bf | FLAKY         | 1/3      |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_dropped_column_no_collision.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.produce_mode=ProducerType.PROTO3.catalog_type=CatalogType.NESSIE    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62708#01956a1e-f1f4-4c59-b704-4cec7e4cfdc2 | FLAKY         | 1/12     |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_dropped_column_no_collision.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.produce_mode=ProducerType.PROTO3.catalog_type=CatalogType.REST_JDBC | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62708#01956a1b-fc0e-4fb4-8a2d-c0dd68f992ca | FLAKY         | 1/4      |
| rptest.tests.datalake.schema_evolution_test.SchemaEvolutionE2ETests.test_dropped_column_no_collision.cloud_storage_type=CloudStorageType.S3.query_engine=QueryEngineType.TRINO.produce_mode=ProducerType.PROTO3.catalog_type=CatalogType.REST_JDBC | ducktape    | https://buildkite.com/redpanda/redpanda/builds/62708#01956a1e-f1f6-4c2c-996f-679eac581f9d | FLAKY         | 1/2      |
</details>
<!--/body-->
